### PR TITLE
server: return non-live, live info per table

### PR DESCRIPTION
**Pull request by maryliag**
This commit adds 3 new parameter to the table
details endpoint:
- dataTotalBytes
- dataNonLiveBytes
- dataNonLivePercentage

Partially addresses #82617

Release note (api change): Add information about total bytes,
non live (MVCC) bytes and non live (MVCC) percentage to
Table Details endpoint.

### CODE REVIEW

**Comment by maryliag** [comment]
@ajwerner I create this version with the most straight forward solution. Are there improvements you can see I should be adding here from the start?

File: pkg/server/admin.go
Comment (ajwerner) [inline pkg/server/admin.go:909]: I don't think you need to quote this. In fact, I'm surprised that this works with quotes around it like this.

**Comment by ajwerner** [comment]
My fear is that for large tables this is going to now make the request just not work where before it worked. I don't really know what to do about that.

File: pkg/server/admin.go
Comment (maryliag) [reply pkg/server/admin.go]: Fixed

**Comment by maryliag** [comment]
Anything we can do with the info about total range count from [here](https://github.com/cockroachdb/cockroach/blob/1f66aa0ab05f2822e9c485a7af91877b7d3d0982/pkg/server/admin.go#L1024) which is just a little below (and I could move my changes to after that)?
In the sense of if this count passes a threshold, I don't try to calculate the garbage? Or have another endpoint specific for it which will load the value independently? But if we follow your example of waiting 8min for the result, might not be worth trying this at all (at least with this implementation)

**Comment by ajwerner** [comment]
I have a cute idea which might solve both the problems here and the problems in `crdb_internal.ranges`. What if we built a vectorized operator for the function? Then we could issue the range status lookups in parallel. It might even be backportable to fix the `crdb_internal.ranges` issues.

@yuzefovich how crazy does it sound? It sort of looks like if I had an implementation and hooked it up in the below code, then maybe it'd work? Is there more to know about planning to make the pieces fit together?

https://github.com/cockroachdb/cockroach/blob/928f605c3e84efb30cab213b47b2f4a66b816701/pkg/sql/colexec/builtin_funcs.go#L122

**Review by yuzefovich — COMMENTED** [review]
I'm not sure I follow your idea - what exactly would benefit from the vectorization? Or are you thinking that in the vectorized infrastructure we could create a specialized operator for a particular builtin function which internally has parallelism (to issue those range status lookups) unlike in the row-by-row infra? That seems reasonable to me, and I think the code you're pointing at would be the right entry point, and then things should mostly just work.

There would be some minor changes to expose the fact of concurrency in this builtin (if we're using a Txn, we'd have to ask for a LeafTxn, also `vectorizedFlowCreator.operatorConcurrency` from `vectorized_flow.go` will need to be set to `true` - actually doing that will ensure that we'll get a LeafTxn).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83677#-:-N5rdsdy8-EPhMTJg3E8:b-bac7yj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

**Comment by ajwerner** [comment]
>  Or are you thinking that in the vectorized infrastructure we could create a specialized operator for a particular builtin function which internally has parallelism (to issue those range status lookups) unlike in the row-by-row infra? 

Precisely.

> There would be some minor changes to expose the fact of concurrency in this builtin (if we're using a Txn, we'd have to ask for a LeafTxn, also vectorizedFlowCreator.operatorConcurrency from vectorized_flow.go will need to be set to true - actually doing that will ensure that we'll get a LeafTxn).

Fortunately `RangeStatsRequest` is non-transactional, so I don't think we'd need anything like that. Any other gotchas?

**Comment by yuzefovich** [comment]
> Any other gotchas?

Hm, no, nothing else comes to mind (as long as the queries using this new builtin don't also include things like CDC / BulkIO - those don't run through the vectorized infra).

**Comment by ajwerner** [comment]
@yuzefovich maybe you can help me out. I typed the code but it doesn't quite work (also I'm sure I'm not doing memory monitoring correctly, or at all). My understanding of why is that we don't actually vectorize the rendering of the expression. What can I do to force it to use the vectorized operator?

```
explain (vec) select crdb_internal.range_stats(start_key) from crdb_internal.ranges limit 1;
              info
--------------------------------
  │
  └ Node 1
    └ *sql.planNodeToRowSource
(3 rows)
```

leads to the builtin being called as part of evaluation in the `ProcessRowHelper` as part of `planNodeToRowSource`.

**Comment by ajwerner** [comment]
Maybe this code here is the problem, or at least, the source of inflexibility: https://github.com/cockroachdb/cockroach/blob/928f605c3e84efb30cab213b47b2f4a66b816701/pkg/sql/distsql_spec_exec_factory.go#L168-L176

**Comment by yuzefovich** [comment]
I think the problem is the usage of `crdb_internal.ranges` which is a materialized view for which we don't have a corresponding processor / operator, so it gets wrapped in https://github.com/cockroachdb/cockroach/blob/78caa5c3a8ecd0d0a612798a106b626edb3db2ee/pkg/sql/distsql_physical_planner.go#L3083 (essentially the same code that you linked to, just on the main exec factory path).

**Comment by yuzefovich** [comment]
(Maybe it's not a "materialized view" but it's something (namely "virtual table" inside of it - `crdb_internal.ranges_no_leases`) that is not powered by the DistSQL, so we, instead, create a `planNode` that is wrapped via `LocalPlanNode` wrapper into the DistSQL infra). (This could be more easily seen with `EXPLAIN (DISTSQL)`.) And then you're right - we don't look inside of that `planNode` at all.

**Comment by ajwerner** [comment]
Hmm, is there a way to force the code to look inside the expressions and inject some intermediate render node or something like that if there are optimized functions?  I imagine there's a way to go from rows->columns->render->rows

**Comment by yuzefovich** [comment]
No, unfortunately nothing like that exists. The thing is that this `planNode`-into-DistSQL wrapping was added before the vectorized execution, when we were merging the local and the DistSQL row-by-row execution models, and we never really changed this wrapping. It does try to find the closest child `planNode` that we don't have to wrap, but this is not enough in this case. I think the problem here is that we have `virtualPlanNode` (which internally does the rendering you're interested in optimizing), and we treat this `planNode` as a black-box when performing DistSQL / vectorized planning.

**Comment by yuzefovich** [comment]
Oh, actually, I take this, at least partially, back - the problem might be simpler than I initially thought. It does look like the rendering is in `PostProcessSpec` and not in the processor core, and we could do the vectorized planning for it. I actually have been thinking about doing that for some time (in order to support more efficient vectorized filtering on top of the `joinReader` processor). So you are right - we should be able to get to "rows -> columns -> vectorized rendering -> rows" model.

**Comment by yuzefovich** [comment]
I typed up #83689 for that idea, but I think it's broken. However, it might be sufficient for your prototype.

**Comment by ajwerner** [comment]
I didn't really follow your draft PR. Maybe we can collaborate on this for a little bit this afternoon if you're willing or some time next week. I've reached the point in the code where my lack of understanding of concepts means I'd need to learn a lot to make progress.

**Comment by yuzefovich** [comment]
I polished #83689, and I think it is in a good shape (modulo green CI), so @ajwerner if you try your prototype with the vectorized builtin on top of that PR, I think you should see the builtin be vectorized.

**Comment by yuzefovich** [comment]
(Bart delays provide ample time to code lol)

**Comment by yuzefovich** [comment]
Indeed:
```
root@127.0.0.1:26257/defaultdb> explain (vec) select crdb_internal.range_stats(start_key) from crdb_internal.ranges limit 1;
                   info
-------------------------------------------
  │
  └ Node 1
    └ *colexec.defaultBuiltinFuncOperator
      └ *sql.planNodeToRowSource
(4 rows)
```

**Comment by ajwerner** [comment]
Your code seems sad here: https://github.com/cockroachdb/cockroach/blob/8b83a216dc4210cea919dc54016a67208a0a2bdf/pkg/server/admin.go#L933

It seems that the type is wrong the column it claims. Maybe you need to name the projection with the cast? I'm not sure.

**Comment by ajwerner** [comment]
```
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600  source type assertion failed                                                                                                                                                                                                                  
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +(1) attached stack trace                                                                                                                                                                                                                      
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  -- stack trace:                                                        
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/server.resultScanner.ScanIndex                                                                                                                                                                       
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/server/admin.go:3159                                                                                                                                                                         
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/server.resultScanner.Scan                                                                                                                                                                            
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/server/admin.go:3261                                                                                                                                                                         
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/server.(*adminServer).tableDetailsHelper                                                                                                                                                             
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/server/admin.go:933                                                                                                                                                                          
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/server.(*adminServer).TableDetails                                                                                                                                                                   
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/server/admin.go:678                                                                                                                                                                          
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/server/serverpb._Admin_TableDetails_Handler.func1                                                                                                                                                    
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/server/serverpb/bazel-out/darwin-fastbuild/bin/pkg/server/serverpb/serverpb_go_proto_/github.com/cockroachdb/cockroach/pkg/server/serverpb/admin.pb.go:4817                                  
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ServerInterceptor.func1                                                                                                                                                 
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:114                                                                                                                                         
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | google.golang.org/grpc.chainUnaryInterceptors.func1.1                                                                                                                                                                                     
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         google.golang.org/grpc/external/org_golang_google_grpc/server.go:1117                                                                                                                                                             
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3                                                                                                                                                                                
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:264                                                                                                                                                                   
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | google.golang.org/grpc.chainUnaryInterceptors.func1.1                                                                                                                                                                                     
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         google.golang.org/grpc/external/org_golang_google_grpc/server.go:1120                                                                                                                                                             
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor                                                                                                                                                                          
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:73                 
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | google.golang.org/grpc.chainUnaryInterceptors.func1.1                           
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         google.golang.org/grpc/external/org_golang_google_grpc/server.go:1120                                                                                                                                                             
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1                                                                                                                                                                              
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:233                                                                                                                                                                   
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr                                                                                                                                                                  
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:341                                                                                                                                                                     
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1                                                                                                                                                                                
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:231                                                                                                                                                                   
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | google.golang.org/grpc.chainUnaryInterceptors.func1.1                                
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         google.golang.org/grpc/external/org_golang_google_grpc/server.go:1120
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | google.golang.org/grpc.chainUnaryInterceptors.func1                                                                                                                                                                                       
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         google.golang.org/grpc/external/org_golang_google_grpc/server.go:1122                                                                                                                                                             
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | github.com/cockroachdb/cockroach/pkg/server/serverpb._Admin_TableDetails_Handler                                                                                                                                                          
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         github.com/cockroachdb/cockroach/pkg/server/serverpb/bazel-out/darwin-fastbuild/bin/pkg/server/serverpb/serverpb_go_proto_/github.com/cockroachdb/cockroach/pkg/server/serverpb/admin.pb.go:4819                                  
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | google.golang.org/grpc.(*Server).processUnaryRPC                                                                                                                                                                                          
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         google.golang.org/grpc/external/org_golang_google_grpc/server.go:1283                                                                                                                                                             
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | google.golang.org/grpc.(*Server).handleStream                                                                                                                                                                                             
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         google.golang.org/grpc/external/org_golang_google_grpc/server.go:1620                                                                                                                                                             
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | google.golang.org/grpc.(*Server).serveStreams.func1.2                                                                                                                                                                                     
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         google.golang.org/grpc/external/org_golang_google_grpc/server.go:922                                                                                                                                                              
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  | runtime.goexit                                                                                                                                                                                                                            
E220707 14:03:18.840447 5147 server/admin.go:300  [n1] 600 +  |         GOROOT/src/runtime/asm_amd64.s:1581 
```

https://github.com/cockroachdb/cockroach/pull/83992

**Review by matthewtodd — APPROVED** [review]
LGTM pending #83935

**Comment by maryliag** [comment]
TFTR!
bors r+

**Pull request merged**
